source/mir/random/discrete.d

```diff
@@ -11,6 +11,221 @@ module mir.random.discrete;
+    auto ds = discrete(probs);
+
+    // sample from the discrete distribution
+    auto obs = new uint[probs.length];
+    foreach (i; 0..10_000)
+        obs[ds()]++;
+}
+
+/**
+_Discrete distribution sampler that draws random values from a _discrete
+distribution given an array of the respective probabilities.
+
+Complexity: O(1) per sampling and O(n) for the initial setup, where n is the number
+            of discrete values.
+
+References:
+    Vose, Michael D. "A linear algorithm for generating random numbers with a given distribution."
+    IEEE Transactions on software engineering 17.9 (1991): 972-975.
+*/
+struct Discrete(T)
+    if (isNumeric!T)
+{
+
+    /// array with the original column value for a discrete value an
+    static struct AltPair
+    {
+        T prob; /// probability p to select it by a coin toss, if this column is randomly picked
+        size_t alt; /// alternative value if coin toss at j fails
+    }
+
+    private AltPair[] arr;
+
+    /**
+    Initialize a discrete distribution sampler
+    Params:
+        probs = probabilities of the individual, discrete values
+
+    Complexity: O(n), where n is the number of discrete values
+    */
+    this(const(T)[] probs)
+    {
+        debug
+        {
+            import mir.sum : sum, Summation;
+            import std.math : approxEqual;
+            assert(probs.sum!(Summation.fast).approxEqual(1.0), "Sum of the probabilities must be 1");
+        }
+        initialize(probs);
+    }
+
+    /**
+    Initialize procedure after [Vose91].
+    Each column of the probability array is either filled or used to fill a column
+    in such a order that each column has exactly two indexes - the original and
+    the alias.
+    This procedure is stable and guaranteed to find such columns as the probabilities
+    are split at the average probability p_avg and each column is picked with
+    1/n - the average probability, so larger columns that are used to fill always
+    have enough capacity.
+    Afterwards in the sampling step such a column is picked by rolling a fair dice (1/n),
+    and whether the original or alias entry in the column should be used is picked
+    by tossing a coin with the probability equal to `p_column * n`.
+    A detailed proof can be found at [Vose91].
+    */
+    private void initialize(const(T)[] probs)
+    {
+        size_t n = probs.length;
+        arr = new AltPair[probs.length];
+
+        static struct ProbsIndexed
+        {
+            T prob; // scaled probability
+            size_t index; // original column index
+        }
+
+        size_t smCounter, lgCounter;
+
+        import std.experimental.allocator.mallocator : Mallocator;
+        import std.experimental.allocator : dispose, makeArray;
+        auto alloc = Mallocator.instance;
+
+        auto small = alloc.makeArray!ProbsIndexed(n); // columns that need to be filled
+        auto large = alloc.makeArray!ProbsIndexed(n); // used to fill columns
```

Comment:
One array should be enough. It would be a little bit more cache friendly, and require only one system call. The idea is use the end of array for reversed stack.